ksql-serde/src/main/java/io/confluent/ksql/serde/connect/ConnectDataTranslator.java

```diff
@@ -0,0 +1,121 @@
+    }
+    final Struct rowStruct = (Struct) toKsqlValue(schema, connectSchema, connectData);
+    return new GenericRow(
+        schema.fields()
+            .stream()
+            .map(f -> rowStruct.get(f.name()))
+            .collect(Collectors.toList())
+    );
+  }
+
+  private Object toKsqlValue(final Schema schema, final Schema connectSchema,
+                             final Object connectValue) {
+    // Map a connect value+schema onto the schema expected by KSQL. For now this involves:
+    // - handling case insensitivity for struct field names
+    // - setting missing values to null
+
+    if (!schema.type().equals(connectSchema.type())) {
+      throw new DataException(
+          String.format(
+              "Cannot deserialize type %s as type %s",
+              connectSchema.type().getName(),
+              schema.type().getName()));
+    }
+    if (connectValue == null) {
+      return null;
+    }
+    if (schema.type().isPrimitive()) {
+      return connectValue;
+    }
+    if (schema.type().equals(Schema.Type.ARRAY)) {
+      return toKsqlArray(schema.valueSchema(), connectSchema.valueSchema(), (List) connectValue);
+    }
+    if (schema.type().equals(Schema.Type.MAP)) {
+      return toKsqlMap(
+          schema.keySchema(), connectSchema.keySchema(),
+          schema.valueSchema(), connectSchema.valueSchema(), (Map) connectValue);
+    }
+    return toKsqlStruct(schema, connectSchema, (Struct) connectValue);
+  }
+
+  private List toKsqlArray(final Schema valueSchema, final Schema connectValueSchema,
+                           final List<Object> connectArray) {
+    return connectArray.stream()
+        .map(o -> toKsqlValue(valueSchema, connectValueSchema, o))
+        .collect(Collectors.toList());
+  }
+
+  private Map toKsqlMap(final Schema keySchema, final Schema connectKeySchema,
+                        final Schema valueSchema, final Schema connectValueSchema,
+                        final Map<String, Object> connectMap) {
+    return connectMap.entrySet().stream()
+        .collect(
+            Collectors.toMap(
+                e -> toKsqlValue(keySchema, connectKeySchema, e.getKey()),
+                e -> toKsqlValue(valueSchema, connectValueSchema, e.getValue())
+            ));
+  }
+
+  private Struct toKsqlStruct(final Schema schema, final Schema connectSchema,
+                              final Struct connectStruct) {
+    // todo: check name here? e.g. what if the struct gets changed to a union?
+    final Struct ksqlStruct = new Struct(schema);
+    final Map<String, String> caseInsensitiveFieldNameMap
+        = getCaseInsensitiveFieldMap(connectStruct.schema());
+    for (Field field : schema.fields()) {
+      // TODO: should we throw an exception if this is not true? this means the schema changed
+      //       or the user declared the source with a schema incompatible with the registry schema
+      if (caseInsensitiveFieldNameMap.containsKey(field.name().toUpperCase())) {
```

Comment:
Maybe define a variable for `field.name().toUpperCase()`. You are using it 3 times here.